### PR TITLE
Homework12: Valeria Kudelko

### PR DESCRIPTION
**Pull request by ValeriaKudelko**

### CODE REVIEW

File: hw12task1.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+
+import random
+
+
+class Card:
+    """
+    Contains a list of card numbers and a list of suits.
+    """
+    number_list = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack', 'Queen', 'King', 'Ace']  # pylint: disable=line-too-long  # noqa: E501
```
Comment (Stml89) [inline hw12task1.py]: It should be in .pylintrc or pylint workflow file
```suggestion
    number_list = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack', 'Queen', 'King', 'Ace']
```

File: hw12task1.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        """
+        random.shuffle(self.deck)
+
+    def get(self, card_number):
+        """
+        Output card.
+        """
+        if 1 <= card_number <= len(self.deck):
+            return self.deck[card_number - 1]
+        return "Joker"
```
Comment (Stml89) [inline hw12task1.py]: Add 2 Jokers to the deck as you did for the other cards.

File: hw12task3.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    matches = re.findall(pattern, formula)
+    atoms = {}
+    for match in matches:
+        atom = match[0]
+        count = int(match[1]) if match[1] else 1
+        atoms[atom] = atoms.get(atom, 0) + count
+    return atoms
+
+
+formulas = ["H2O", "Mg(OH)2", "K4[ON(SO3)2]2"]
```
Comment (Stml89) [inline hw12task3.py]: ```K4[ON(SO3)2]2 -> {'K': 4, 'O': 4, 'N': 1, 'S': 1}``` but should be ``` {K: 4, O: 14, N: 2, S: 4} ```

File: hw12task2.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        self.currency_converter = Currency()
+
+    def exchange_currency(self, source_currency, amount, target_currency=None):
+        """
+        Converts the given amount from the source
+        currency to the target currency, or to BYN if
+        no target currency is specified.
+        """
+        if not target_currency:
+            target_currency = 'BYN'
```
Comment (Stml89) [inline hw12task2.py]:
```suggestion
    def exchange_currency(self, source_currency, amount, target_currency='BYN'):
        """
        Converts the given amount from the source
        currency to the target currency, or to BYN if
        no target currency is specified.
        """
```

**Review by Stml89 — CHANGES_REQUESTED** [review]

File: hw12task1.py (hunk 1 above)
Comment (ValeriaKudelko) [reply hw12task1.py]: Removed

File: hw12task1.py (hunk 2 above)
Comment (ValeriaKudelko) [reply hw12task1.py]: Added

File: hw12task3.py (hunk 3 above)
Comment (ValeriaKudelko) [reply hw12task3.py]: Corrected

**Comment by ValeriaKudelko** [comment]
Test close

File: hw12task1.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+import random
+
+
+class Card:
+    """
+    Contains a list of card numbers and a list of suits.
+    """
+    number_list = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack',
+                   'Queen', 'King', 'Ace', 'Joker']
```
Comment (Stml89) [inline hw12task1.py:41]:
```suggestion
                   'Queen', 'King', 'Ace']
```

File: hw12task1.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    def __str__(self):
+        return f"{self.mast} {self.number}"
+
+
+class CardsDeck:
+    """
+    Contains an entire deck of cards.
+    """
+    def __init__(self):
+        self.deck = [Card(number, mast) for number in Card.number_list for mast in Card.mast_list]  # noqa: 501
```
Comment (Stml89) [inline hw12task1.py:57]:
```suggestion
        self.deck = [Card(number, mast) for number in Card.number_list for mast in Card.mast_list]
```

File: hw12task3.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            current_group = {}
+            i += 1
+        elif char in (')', ']', '}'):
+            factor = int(formula[i + 1]) if i + 1 < len(formula) and formula[i + 1].isdigit() else 1  # noqa: 501
+            current_group = multiply_dict(current_group, factor)
+            prev_group = stack.pop()
+            current_group = merge_dicts(prev_group, current_group)
+            i += 2
+        else:
+            group_match = re.match(r'([A-Z][a-z]*)(\d*)', formula[i:])
```
Comment (Stml89) [inline hw12task3.py:66]: the same formula in line #43
maybe it is a good idea to move formula to a variable:
formula = "([A-Z][a-z]*)(\d*)"

**Review by Stml89 — CHANGES_REQUESTED** [review]

**Pull request merged**